librad/tests/propagation_basic.rs

```diff
@@ -331,3 +331,64 @@ async fn all_metadata_returns_only_local_projects() {
     })
     .await;
 }
+
+/// Given that a) a peer 1 holds a given URN and b) that same peer is a seed of
+/// a peer 2, verify that requesting peer 2 for providers for said URN returns
+/// peer 1.
+#[tokio::test]
+async fn providers_works() {
+    use std::time::Duration;
+    use tokio::time::timeout;
+
+    logging::init();
+    const NUM_PEERS: usize = 2;
+    let peers = testnet::setup(NUM_PEERS).await.unwrap();
+
+    testnet::run_on_testnet(peers, NUM_PEERS, async move |mut apis| {
+        let (peer1, peer1_key) = apis.pop().unwrap();
+        let mut alice = Alice::new(peer1_key.public());
+        let mut radicle = Radicle::new(&alice);
+        let resolves_to_alice = alice.clone();
+        alice
+            .sign(&peer1_key, &Signatory::OwnedKey, &resolves_to_alice)
+            .unwrap();
+        radicle
+            .sign(
+                &peer1_key,
+                &Signatory::User(alice.urn()),
+                &resolves_to_alice,
+            )
+            .unwrap();
+
+        let peer1_id = peer1.peer_id().clone();
+        let repo_urn = tokio::task::spawn_blocking(move || {
```

Comment:
Good find :+1: This blocking IO stuff is quite the pain

Reply:
kudus to the one and only, @xla

Reply:
Why is this necessary? `create_repo` is completely sync.

Reply:
I think the issue is that `create_repo` _is_ sync, i.e, blocking. Looking at the docs of `tokio::task::spawn_blocking`: 

> In general, issuing a blocking call or performing a lot of compute in a future without yielding is not okay, as it may prevent the executor from driving other futures forward. This function runs the provided closure on a thread dedicated to blocking operations.

Or did I misunderstood the issue, @xla?

Reply:
I was a bit sceptical too and tried running it without the spawn blocking. The call to `providers` fails.

Is there some asynchrony between the storage and receiving gossip requests maybe?

Reply:
The issue wasn't quite around receiving gossip requests, it was that we were querying the network for the given URN before the seed had connected.

Reply:
Sure, but you don't rely on other futures being polled while this block is executed. This is different from `fetch_repo`, f.ex., which is not _truly_ sync, as it has to drive the network stack's executor by means of going through the (sync) FFI. Note that it does not help to mark that method `async` -- if it is not spawned via a blocking-aware spawner, nothing will happen.

Async Rust is just lovely.

Reply:
> The issue wasn't quite around receiving gossip requests, it was that we were querying the network for the given URN before the seed had connected.

But the `testnet` setup code waits for the seeds to be connected to each other. So that seems weird to me.

What I meant by my gossip comment was that this block indicates to me that we needed to ensure that the project was created in `peer1` before `peer2` made the query asking for the URN. Indicating further that `peer1` had to ensure it had the project before it _received_ a gossip query to check it did indeed have the URN requested.

So my confusion as to why it's needed persists tbh...

Reply:
@FintanH thanks for the clarification, I understand now (also understand something @xla has said before).

> So my confusion as to why it's needed persists tbh...

indeed...

Reply:
Oooh, so what happens here is this:

The test harness waits until `n` peer connections have been established, but it doesn't wait for the gossip membership to settle. By spawning the create repo stuff, we accidentally give it enough time to do that. If we don't, asking the other end for something wins the race (but the other end is not "there" yet).

So tl;dr we need a way to make the test harness wait on `Join` completion, otherwise our tests only work accidentally.

Reply:
Are we able to inspect those types of `ProtocolEvent`s currently?

Reply:
Can confirm it's a matter of delaying after trying out the test by using `tokio::time::delay_for(Duration::from_secs(1)).await;` instead.

Reply:
#313